File: src/test/java/uk/gov/justice/hmpps/prison/service/OffenderEducationServiceTest.java
package uk.gov.justice.hmpps.prison.service;

import java.util.Collections;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import uk.gov.justice.hmpps.prison.repository.jpa.model.Country;
import uk.gov.justice.hmpps.prison.repository.jpa.model.EducationLevel;
import uk.gov.justice.hmpps.prison.repository.jpa.model.EducationSchedule;
import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderEducation;
import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderEducation.PK;
import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderEducationAddress;
import uk.gov.justice.hmpps.prison.repository.jpa.model.StudyArea;
import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderEducationRepository;
import uk.gov.justice.hmpps.prison.service.transformers.OffenderEducationTransformer;

import java.time.LocalDate;
import java.util.List;

import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class OffenderEducationServiceTest {

    private final int BATCH_SIZE = 1;

    private final String nomisId = "abc";

    private final PageRequest pageRequest = PageRequest.of(0, 10);

    private final OffenderEducationAddress address = OffenderEducationAddress.builder()
        .addressId(2349706L)
        .noFixedAddressFlag("N")
        .primaryFlag("N")
        .country(new Country("ENG", "England"))
        .build();


    final OffenderEducation education1 = OffenderEducation
        .builder()
        .id(new PK(584215L, 1L))
        .startDate(LocalDate.of(2009, 12, 21))
        .studyArea(new StudyArea("GEN", "General Studies"))
        .educationLevel(new EducationLevel("DEGREE", "Degree Level or Higher"))
        .comment("Good student")
        .school("School of economics")
        .isSpecialEducation(true)
        .schedule(new EducationSchedule("PART", "Part Time"))
        .addresses(List.of(address))
        .build();

    private final OffenderEducation education2 = OffenderEducation
        .builder()
        .id(new PK(584215L, 2L))
        .startDate(LocalDate.of(2016, 2, 10))
        .comment("Needs more focus")
        .school("moj education")
        .schedule(new EducationSchedule("NK", "Not Known"))
        .isSpecialEducation(false)
        .build();

    @Mock
    private OffenderEducationTransformer transformer;

    @Mock
    private OffenderEducationRepository repository;


    private OffenderEducationService service;


    @BeforeEach
    void setup() {
        service = new OffenderEducationService(repository, transformer, BATCH_SIZE);
    }

    @Test
    public void getOffenderEducations() {
        final var educations = List.of(education1, education2);

        when(repository.findAllByNomisId(nomisId, pageRequest)).thenReturn(new PageImpl<>(educations));

        service.getOffenderEducations(nomisId, pageRequest);

        verify(transformer, times(1)).convert(education1);
        verify(transformer, times(1)).convert(education2);
    }

    @Test
    public void getOffenderEducationsInBulk_inBatchesOfOne() {
        final var educations = List.of(education1, education2);

        when(repository.findAllByNomisIdIn(List.of(nomisId))).thenReturn(educations);
        when(repository.findAllByNomisIdIn(List.of("ABC123"))).thenReturn(Collections.emptyList());

        service.getOffenderEducations(List.of(nomisId, "ABC123"));

        verify(repository).findAllByNomisIdIn(List.of(nomisId));
        verify(repository).findAllByNomisIdIn(List.of("ABC123"));
        verify(transformer, times(1)).convert(education1);
        verify(transformer, times(1)).convert(education2);
    }
}

File: src/test/java/uk/gov/justice/hmpps/prison/service/OffenderEducationServiceTest.kt
package uk.gov.justice.hmpps.prison.service

import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import uk.gov.justice.hmpps.prison.repository.jpa.model.Country
import uk.gov.justice.hmpps.prison.repository.jpa.model.EducationLevel
import uk.gov.justice.hmpps.prison.repository.jpa.model.EducationSchedule
import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderEducation
import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderEducationAddress
import uk.gov.justice.hmpps.prison.repository.jpa.model.StudyArea
import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderEducationRepository
import uk.gov.justice.hmpps.prison.service.transformers.OffenderEducationTransformer
import java.time.LocalDate

class OffenderEducationServiceTest {
  private companion object {
    private const val BATCH_SIZE = 1
    private const val NOMIS_ID = "abc"
  }

  private val address: OffenderEducationAddress = OffenderEducationAddress.builder()
    .addressId(2349706L)
    .noFixedAddressFlag("N")
    .primaryFlag("N")
    .country(Country("ENG", "England"))
    .build()

  val education1: OffenderEducation = OffenderEducation
    .builder()
    .id(OffenderEducation.PK(584215L, 1L))
    .startDate(LocalDate.of(2009, 12, 21))
    .studyArea(StudyArea("GEN", "General Studies"))
    .educationLevel(EducationLevel("DEGREE", "Degree Level or Higher"))
    .comment("Good student")
    .school("School of economics")
    .isSpecialEducation(true)
    .schedule(EducationSchedule("PART", "Part Time"))
    .addresses(listOf(address))
    .build()

  private val education2: OffenderEducation = OffenderEducation
    .builder()
    .id(OffenderEducation.PK(584215L, 2L))
    .startDate(LocalDate.of(2016, 2, 10))
    .comment("Needs more focus")
    .school("moj education")
    .schedule(EducationSchedule("NK", "Not Known"))
    .isSpecialEducation(false)
    .build()

  private val transformer: OffenderEducationTransformer = mock()
  private val repository: OffenderEducationRepository = mock()

  private val service = OffenderEducationService(repository, transformer, BATCH_SIZE)

  @Test
  fun getOffenderEducationsInBulk_inBatchesOfOne() {
    val educations = listOf(education1, education2)

    whenever(repository.findAllByNomisIdIn(listOf(NOMIS_ID)))
      .thenReturn(educations)
    whenever(repository.findAllByNomisIdIn(mutableListOf("ABC123")))
      .thenReturn(
        mutableListOf<OffenderEducation?>(),
      )

    service.getOffenderEducations(listOf(NOMIS_ID, "ABC123"))

    verify(repository).findAllByNomisIdIn(listOf(NOMIS_ID))
    verify(repository).findAllByNomisIdIn(mutableListOf("ABC123"))
    verify(transformer).convert(education1)
    verify(transformer).convert(education2)
  }
}
